tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/Precision.java

```diff
@@ -0,0 +1,419 @@
+      long seed,
+      Class<T> type) {
+    super(tf, name, seed);
+    this.type = type;
+    this.truePositivesName = this.getVariableName(TRUE_POSITIVES);
+    this.falsePositivesName = this.getVariableName(FALSE_POSITIVES);
+    float defaultThreshold = topK == null ? DEFAULT_THRESHOLD : MetricsHelper.NEG_INF;
+    this.thresholds = thresholds == null ? new float[] {defaultThreshold} : thresholds;
+    this.topK = topK;
+    this.classId = classId;
+
+    init();
+  }
+
+  /** Initializes the variables */
+  private void init() {
+    Ops tf = getTF();
+    Zeros<T> zeros = new Zeros<>(tf);
+    Operand<T> zero = zeros.call(tf.constant(Shape.of(thresholds.length)), type);
+
+    if (this.truePositives == null) {
+      this.truePositives = tf.withName(truePositivesName).variable(zero);
+      initializers.add(tf.assign(truePositives, zero));
+    }
+    if (this.falsePositives == null) {
+      this.falsePositives =
+          tf.withName(falsePositivesName)
+              .variable(zeros.call(tf.constant(Shape.of(thresholds.length)), type));
+      initializers.add(tf.assign(falsePositives, zero));
+    }
+  }
+
+  /**
+   * Accumulates true positive and false positive statistics.
+   *
+   * @param labels the labels The ground truth values, with the same dimensions as predictions. Will
+   *     be cast to {@link TBool}.
+   * @param predictions the predictions, each element must be in the range {@code [0, 1]}.
+   * @param sampleWeights Optional weighting of each example. Defaults to 1. Rank is either 0, or *
+   *     the same rank as labels, and must be broadcastable to labels.
+   * @return a List of Operations to update the metric state.
+   */
+  @Override
+  @SuppressWarnings("unchecked")
+  public List<Op> updateStateList(
+      Operand<? extends TNumber> labels,
+      Operand<? extends TNumber> predictions,
+      Operand<? extends TNumber> sampleWeights) {
+    Ops tf = getTF();
+    Map<ConfusionMatrixEnum, Variable<T>> confusionMatrix = new HashMap<>();
+    confusionMatrix.put(ConfusionMatrixEnum.TRUE_POSITIVES, truePositives);
+    confusionMatrix.put(ConfusionMatrixEnum.FALSE_POSITIVES, falsePositives);
+
+    Operand<T> tPredictions = cast(tf, predictions, type);
+    Operand<T> tLabels = cast(tf, labels, type);
+    Operand<T> tSampleWeights = sampleWeights != null ? cast(tf, sampleWeights, type) : null;
+
+    return new ArrayList<Op>(
+        MetricsHelper.updateConfusionMatrixVariables(
+            tf,
+            confusionMatrix,
+            Collections.EMPTY_MAP,
+            tLabels,
+            tPredictions,
+            tf.constant(thresholds),
+            topK,
+            classId,
+            tSampleWeights,
+            false,
+            null));
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public Operand<T> result() {
+    Ops tf = getTF();
+    Operand<T> result = tf.math.divNoNan(truePositives, tf.math.add(truePositives, falsePositives));
+    return thresholds.length == 1
+        ? tf.slice(
```

Comment:
I don't think this accurately mirrors the Python. Also, I think the fact that our tests don't catch the problem shows a weakness in our test infrastructure.

Here's the Python code:
~~~Python
    return result[0] if len(self.thresholds) == 1 else result
~~~
In the case of a single threshold, it returns a scalar. But I believe this Java code returns shape (1).

Here's how we verify a return value in `PrecisionTest`:
~~~Java
      double expectedPrecision = weightedTP / weightedPositives;
      session.evaluate(expectedPrecision, precision);
~~~
This is intended to verify that `precision` is a scalar with value `expectedPrecision`. But here's the verification code in `org.tensorflow.framework.metrics.PrecisionTest#testWeighted`:
~~~Java
      try (TFloat32 result =
          (TFloat32)this.getGraphSession().runner().fetch(input).run().get(0)) {
        result.scalars().forEach(f -> assertEquals((float) expected, f.getFloat(), epsilon));
      }
~~~
It doesn't verify that `result` is a scalar. Rather, it verifies that every scalar in `result` has the expected value. (In fact, I think it `result` were an empty list, the test pass?)

Reply:
You are right, Python returns scalar, I have change to:
```
return  thresholds.length == 1
        ? tf.reshape(tf.slice(
            result,
            tf.expandDims(tf.constant(0), tf.constant(0)),
            tf.expandDims(tf.constant(1), tf.constant(0))),
            tf.constant(Shape.scalar()))
        : result;
```